### PR TITLE
Adding dualstack overlay support to azure-ipam plugin for Cilium

### PR DESCRIPTION
**Pull request by rjdenney**
**Reason for Change**:

Adds support in the azure-ipam plugin to use new APIs created for CNS to allow for passing multiple IPs on Add and Delete commands which is needed to create dualstack clusters.

Here's where the new APIs were added.
https://github.com/Azure/azure-container-networking/pull/1773

**Issue Fixed**:

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] relevant PR labels added

**Notes**:

### CODE REVIEW

File: cilium/cilium_helm_values.yaml
Comment (camrynl) [inline cilium/cilium_helm_values.yaml:1]: let's leave this change out since we aren't using helm to install cilium on byocni. we'll probably want to configure conditions later on in the cilium config to set the values for ipv6 since we may not always want to create dualstack clusters

File: azure-ipam/ipam.go [hunk 1]
```diff
@@ -130,18 +161,43 @@ func (p *IPAMPlugin) CmdDel(args *cniSkel.CmdArgs) error {
...
+	if err := p.cnsClient.ReleaseIPs(context.TODO(), req); err != nil {
+		// if we fail a request with a 404 error try using the old API
+		if cnscli.IsUnsupportedAPI(err) {
+			p.logger.Error("Failed to release IPs using ReleaseIPs from CNS, going to try ReleaseIPAddress", zap.Error(err), zap.Any("request", req))
+			ipconfigReq, err := ipconfig.CreateIPConfigReq(args)
+			if err != nil {
+				p.logger.Error("Failed to create CNS IP config request", zap.Error(err))
+				return cniTypes.NewError(ErrCreateIPConfigRequest, err.Error(), "failed to create CNS IP config request")
+			}
+			err = p.cnsClient.ReleaseIPAddress(context.TODO(), ipconfigReq)
```
Comment (vipul-21) [inline azure-ipam/ipam.go:182]: Same comment for logs here as well.

File: cilium/cilium_helm_values.yaml
Comment (camrynl) [reply cilium/cilium_helm_values.yaml:1]: used to be used for ci, but we moved away from that method of install after adding our own init containers to the cilium daemonset

File: azure-ipam/ipam.go [hunk 2]
```diff
@@ -69,40 +71,69 @@ func (p *IPAMPlugin) CmdAdd(args *cniSkel.CmdArgs) error {
...
+	req, err := ipconfig.CreateIPConfigsReq(args)
 	if err != nil {
-		p.logger.Error("Failed to create CNS IP config request", zap.Error(err))
-		return cniTypes.NewError(ErrCreateIPConfigRequest, err.Error(), "failed to create CNS IP config request")
+		p.logger.Error("Failed to create CNS IP configs request", zap.Error(err))
+		return cniTypes.NewError(ErrCreateIPConfigRequest, err.Error(), "failed to create CNS IP configs request")
 	}
 	p.logger.Debug("Created CNS IP config request", zap.Any("request", req))
 
 	p.logger.Debug("Making request to CNS")
```
Comment (tamilmani1989) [inline azure-ipam/ipam.go:81]: these logs should be moved where request is made

File: azure-ipam/ipam.go [hunk 3]
```diff
@@ -130,18 +161,43 @@ func (p *IPAMPlugin) CmdDel(args *cniSkel.CmdArgs) error {
...
+
+			if err != nil {
+				if errors.As(err, &connectionErr) {
+					p.logger.Info("Failed to release IP address from CNS due to connection failure, saving to watcher to delete")
+					addErr := fsnotify.AddFile(args.ContainerID, args.ContainerID, watcherPath)
+					if addErr != nil {
+						p.logger.Error("Failed to add file to watcher", zap.String("containerID", args.ContainerID), zap.Error(addErr))
+						return cniTypes.NewError(cniTypes.ErrTryAgainLater, addErr.Error(), fmt.Sprintf("failed to add file to watcher with containerID %s", args.ContainerID))
+					} else {
+						p.logger.Info("File successfully added to watcher directory")
```
Comment (tamilmani1989) [inline azure-ipam/ipam.go:192]: can we make this change as separate PR?

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

**Comment by tamilmani1989** [comment]
@rjdenney i dont see unit tests added.. can you add for new changes you added?

File: cilium/cilium_helm_values.yaml
Comment (rjdenney) [reply cilium/cilium_helm_values.yaml:1]: removed the helm values

**Review by camrynl — APPROVED** [review]

**Review by vipul-21 — APPROVED** [review]

File: azure-ipam/ipam_test.go [hunk 4]
```diff
@@ -227,15 +385,25 @@ func TestCmdDel(t *testing.T) {
...
 		},
 		{
-			name:    "Fail request CNS release IP during CmdDel",
+			name:    "Fail request CNS release using ReleaseIPAddress during CmdDel",
 			args:    buildArgs("failRequestCNSReleaseIPArgs", happyPodArgs, happyNetConfByteArr),
 			wantErr: true,
 		},
+		{
+			name:    "Fail request CNS release during CmdDel",
+			args:    buildArgs("failRequestCNSReleaseIPsArgs", happyPodArgs, happyNetConfByteArr),
```
Comment (tamilmani1989) [inline azure-ipam/ipam_test.go:404]: this case should validate if azure-ipam calls ReleaseIPAddress api if ReleaseIPs fails

**Comment by tamilmani1989** [comment]
@rjdenney @camrynl should have test case to validate if azure-cns returns "unsupported api" error, cns should call old api

File: azure-ipam/ipam.go
[the remark this answers is not in the document]
Comment (rjdenney) [reply azure-ipam/ipam.go:1]: Here's where the new APIs were added.
https://github.com/Azure/azure-container-networking/pull/1773

File: azure-ipam/ipam_test.go (hunk 4 above)
Comment (rjdenney) [reply azure-ipam/ipam_test.go:404]: Right now in the code the only time that we call `ReleaseIPAddress` is if we get a 404 error when we call `ReleaseIPs`. 

With the four tests that exist now the first happy path uses  `ReleaseIPs` and works fine since it is the happy path, second one first calls `ReleaseIPs`, fails with a 404 error, and then calls `ReleaseIPAddress` which passes, third fails does the same as second but fails at `ReleaseIPAddress`, and the fourth fails `ReleaseIPs` with an error that is not 404 so we don't call `ReleaseIPAddress`.

File: azure-ipam/ipam_test.go [hunk 5]
```diff
@@ -85,6 +88,114 @@ func (c *MockCNSClient) RequestIPAddress(ctx context.Context, ipconfig cns.IPCon
...
 
+func (c *MockCNSClient) RequestIPs(ctx context.Context, ipconfig cns.IPConfigsRequest) (*cns.IPConfigsResponse, error) {
+	switch ipconfig.InfraContainerID {
+	case "failRequestCNSArgs":
+		return nil, errFoo
+	case "happyArgsSingle", "failProcessCNSRespSingleIP", "failRequestCNSArgsSingleIP":
+		e := &client.CNSClientError{}
+		e.Code = types.UnsupportedAPI
+		e.Err = errUnsupportedAPI
+		return nil, e
```
Comment (rjdenney) [inline azure-ipam/ipam_test.go:99]: @tamilmani1989 For `RequestIPs` each of these tests will return back the UnsupportedAPI when we call `RequestIPs` which then causes us to test `RequestIPAddress`

**Review by tamilmani1989 — APPROVED** [review]
lgtm
